File: leetcode/289. Game of Life.cpp
class Solution {
public:
    void gameOfLife(vector<vector<int>>& board) {
        int x[] = {-1, -1, -1, 0, 0, 1, 1, 1};
        int y[] = {0, 1, -1, 1, -1, 0, 1, -1};
        for(int i = 0; i < board.size(); ++i){
            for(int j = 0; j < board[i].size(); ++j){
                int lives = 0;
                for(int m = 0; m < 8; ++m){
                    int x0 = x[m] + i;
                    int y0 = y[m] + j;
                    if(x0 < 0 || x0 > board.size() - 1 || y0 < 0 || y0 > board[i].size() - 1){
                        continue;
                    }
                    lives += isAlive(board[x0][y0]);
                }

                if(board[i][j] == 1 && lives < 2){
                    board[i][j] = 2;
                } else if(board[i][j] == 1 && (lives == 2 || lives == 3)){
                    board[i][j] = 3;
                } else if(board[i][j] == 1 && lives > 3){
                    board[i][j] = 4;
                } else if(board[i][j] == 0 && lives == 3) {
                    board[i][j] = 5;
                } else {
                    board[i][j] = 0;
                }
            }
        }
        
        for(int i = 0; i < board.size(); ++i){
            for(int j = 0; j < board[i].size(); ++j){
                if(board[i][j] == 2 || board[i][j] == 4){
                    board[i][j] = 0;
                } else if(board[i][j] == 3 || board[i][j] == 5) {
                    board[i][j] = 1;
                }
            }   
        }
    }
    
private:
    int isAlive(int val){
        return val == 2 || val == 3 || val == 4 || val == 1? 1 : 0;
    }
};

File: leetcode/289. Game of Life.py
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        directions = [(1,0), (-1,0), (0,1), (0,-1), (1,1), (1,-1), (-1,1), (-1,-1)]
        m = len(board)
        n = len(board[0])
        
        def live_neighbours(x, y):
            cnt = 0
            for x0, y0 in directions:
                x1, y1 = x+x0, y+y0
                if x1 < 0 or x1 > m-1 or y1 < 0 or y1 > n-1:
                    continue
                if board[x1][y1] == 1 or board[x1][y1] == 2:
                    cnt += 1
            return cnt
        
        for i in range(m):
            for j in range(n):
                live = live_neighbours(i, j)
                if board[i][j] == 0:
                    if live == 3:
                        board[i][j] = 3
                else:
                    if live != 2 and live != 3:
                        board[i][j] = 2
        
        for i in range(m):
            for j in range(n):
                if board[i][j] == 2:
                    board[i][j] = 0
                elif board[i][j] == 3:
                    board[i][j] = 1
